rangeserver/src/cache/memtabledb.rs

```diff
@@ -0,0 +1,643 @@
+                    None => println!("INFO: GC callback is not set {:?}", self.gc_callback),
+                }
+            }
+            self.memtables[self.mutable_id].entries.insert(entry);
+            self.memtables[self.mutable_id].size_in_bytes += entry_size;
+            self.memtables[self.mutable_id].latest_epoch = epoch;
+            self.latest_epoch = epoch;
+
+            Ok(())
+        }
+}
+
+impl Cache for MemTableDB {
+
+    async fn upsert(
+        &mut self,
+        key: bytes::Bytes,
+        val: bytes::Bytes,
+        epoch: u64,
+        ) -> Result<(), Error> {
+
+            // sanity checks
+            assert!(self.mutable_id >= 0 && self.mutable_id < self.rc_options.num_write_buffers, 
+                "mutable_id is out of range ({})", self.mutable_id);
+                        
+            self.upsert_or_delete(key, epoch, false, Some(val)).await
+    }
+
+    async fn delete(
+        &mut self, 
+        key: bytes::Bytes,
+        epoch: u64,
+    ) -> Result<(), Error> {
+
+        // sanity checks
+        assert!(self.mutable_id >= 0 && self.mutable_id < self.rc_options.num_write_buffers, 
+            "mutable_id is out of range ({})", self.mutable_id);
+        
+        self.upsert_or_delete(key, epoch, true, None).await
+    }
+
+    async fn get(
+        &self, 
+        key: &bytes::Bytes,
+        epoch: Option<u64>,
+    ) -> Result<(bytes::Bytes, u64), Error> {
+
+        // sanity checks
+        assert!(self.mutable_id >= 0 && self.mutable_id < self.rc_options.num_write_buffers, 
+            "mutable_id is out of range ({})", self.mutable_id);
+    
+        let mut epoch_internal = 0;
+        match epoch {
+            Some(e) => {
+                epoch_internal = min(e, self.latest_epoch);
+            },
+            None => {
+                epoch_internal = self.latest_epoch;
+            },
+        }
+
+        for i in 0..self.rc_options.num_write_buffers {
+            let id = (self.mutable_id + self.rc_options.num_write_buffers - i) % self.rc_options.num_write_buffers; // mut_id is usize, should not become negative
+            // let memtable_rg = self.memtables[id].read().unwrap();
+            // let memtable_rg = self.memtables[id].read().await;
+            println!("GET-DBG: id {}, epoch {}, oldest_epoch {}, latest_epoch {}", id, epoch_internal, 
+                self.memtables[id].oldest_epoch, self.memtables[id].latest_epoch);
+            if self.memtables[id].status == MemTableStatus::FLUSHED {
+                break
+            }
+            if self.memtables[id].oldest_epoch <= epoch_internal && epoch_internal <= self.memtables[id].latest_epoch {                
+                match self.memtables[id].entries.iter().find(|&&ref entry| entry.key == key && entry.epoch <= epoch_internal)  {
+                    Some(entry) => { 
+                        println!("Found the entry {:?}", entry);
+                        if entry.deleted {
```

Comment:
Out of curiosity, why keep deleted entries in the cache? It looks like `get` behaves the same whether or not it is deleted or never existed to begin with.

Reply:
Clients can access keys from older epochs. Let's say, client inserts a key in epoch 10 and then deletes the same key in epoch 20.
`upsert(k1, v1, 10)
delete(k1, 20)`
`get(k1, 20)` => should return key not found
`get(k1, 10)` => should return v1

Reply:
Thanks for the example - I got it now. Consider adding the example either in `get` or next to the `deleted` field.